### PR TITLE
Bazel test shebangs

### PR DESCRIPTION
**Pull request by aherrmann**
As suggested by @Profpatsch  in https://github.com/NixOS/nixpkgs/pull/66718#issuecomment-522021975 this PR adds a test to Bazel checking that shebangs have been patched appropriately. 

This test flagged a few more instances of invalid python shebangs which are also fixed by this PR.

###### Motivation for this change

See https://github.com/NixOS/nixpkgs/pull/66718#issuecomment-522021975

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Profpatsch

### CODE REVIEW

File: pkgs/development/tools/build-managers/bazel/shebang-test.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{
+  bazel
+, bazelTest
+, runLocal
+, unzip
+}:
+
+# Tests that all shebangs are patched appropriately.
+# #!/usr/bin/... should be replaced by Nix store references.
+# #!.../bin/env python should be replaced by Nix store reference to the python interpreter.
```
Comment (Profpatsch) [inline pkgs/development/tools/build-managers/bazel/shebang-test.nix]: We should even test for `bin/env` if possible, these all indicate bugs if I’m not mistaken.

File: pkgs/development/tools/build-managers/bazel/default.nix [hunk 2]
```diff
@@ -202,6 +202,7 @@ stdenv.mkDerivation rec {
             # about why to create a subdir for the workspace.
             cp -r ${workspaceDir} wd && chmod u+w wd && cd wd
 
+            BAZEL_EXTRACTED=${be.install_dir}
```
Comment (Profpatsch) [inline pkgs/development/tools/build-managers/bazel/default.nix]: Can you pass it through at the nix level instead? You can pass it to the callPackage of `shebang` instead, then it’s less indirect and we drop to the bash level at the last possible moment.

**Review by Profpatsch — COMMENTED** [review]
Great, this removes another source of easy-to-make and hard-to-debug mistakes.

**Comment by Profpatsch** [comment]
@GrahamcOfBorg build bazel.tests

File: pkgs/development/tools/build-managers/bazel/default.nix (hunk 2 above)
Comment (aherrmann) [reply pkgs/development/tools/build-managers/bazel/default.nix]: Done, PTAL.

File: pkgs/development/tools/build-managers/bazel/shebang-test.nix (hunk 1 above)
Comment (aherrmann) [reply pkgs/development/tools/build-managers/bazel/shebang-test.nix]: Done, it found one more instance of `#!/usr/bin/env bash` that I've now patched to use `customBash`. I've also pulled the `python` patching underneath `devtools` into the loop that already existed further down in the `postPatch` script.

**Review by aherrmann — COMMENTED** [review]
@Profpatsch Thanks for the review, I've addressed your comments, PTAL.

File: pkgs/development/tools/build-managers/bazel/default.nix [hunk 3]
```diff
@@ -321,10 +322,11 @@ stdenv.mkDerivation rec {
     '';
 
     genericPatches = ''
-      # Substitute python's stub shebang to plain python path. (see TODO add pr URL)
+      # Substitute j2objc and objc wrapper's python shebang to plain python path.
       # See also `postFixup` where python is added to $out/nix-support
```
Comment (flokli) [inline pkgs/development/tools/build-managers/bazel/default.nix]: both python2 and 3 are added to $out/nix-support

File: pkgs/development/tools/build-managers/bazel/default.nix [hunk 4]
```diff
@@ -321,10 +322,11 @@ stdenv.mkDerivation rec {
     '';
 
     genericPatches = ''
-      # Substitute python's stub shebang to plain python path. (see TODO add pr URL)
+      # Substitute j2objc and objc wrapper's python shebang to plain python path.
```
Comment (flokli) [inline pkgs/development/tools/build-managers/bazel/default.nix]: we should explain why we replace python2 stuff here, but python3 further down

Comment (aherrmann) [reply pkgs/development/tools/build-managers/bazel/default.nix]: I've added a comment explaining this.

File: pkgs/development/tools/build-managers/bazel/default.nix (hunk 3 above)
Comment (aherrmann) [reply pkgs/development/tools/build-managers/bazel/default.nix]: I've made the comments more precise about the Python versions.

**Comment by kalbasit** [comment]
@GrahamcOfBorg build bazel bazel.tests

If test pass, I will merge so we can assert #67835 before merging.

**Comment by Profpatsch** [comment]
@GrahamcOfBorg build bazel.tests

**Comment by Profpatsch** [comment]
The Darwin build has lots of instances of this error:

> ERROR: /private/tmp/nix-build-bazel-test-cpp.drv-1/wd/BUILD:1:1: Target '//:ProjectRunner' depends on toolchain '@local_config_cc//:cc-compiler-darwin', which cannot be found: error loading package '@local_config_cc//': Unable to find package for @rules_cc//cc:defs.bzl: The repository '@rules_cc' could not be resolved.'

**Comment by aherrmann** [comment]
@Profpatsch also

> ```
> ERROR: /private/tmp/nix-build-bazel-test-builtin-rules.drv-0/wd/python/BUILD.bazel:6:1: Target '//python:bin' depends on toolchain '@local_config_cc//:cc-compiler-darwin', which cannot be found: error loading package '@local_config_cc//': Unable to find package for @rules_cc//cc:defs.bzl: The repository '@rules_cc' could not be resolved.'
> ERROR: Analysis of target '//python:bin' failed; build aborted: error loading package '@local_config_cc//': Unable to find package for @rules_cc//cc:defs.bzl: The repository '@rules_cc' could not be resolved.
> INFO: Elapsed time: 9.871s
> INFO: 0 processes.
> FAILED: Build did NOT complete successfully (12 packages loaded, 27 targets configured)
> ERROR: Build failed. Not running target
> ```

That's strange, this PR doesn't touch `@local_config_cc` or `@rules_cc`. Do we know that this did not already happen before this PR?

**Comment by Profpatsch** [comment]
> Do we know that this did not already happen before this PR?

Good question, here’s a dummy PR to test: https://github.com/NixOS/nixpkgs/pull/67999

**Comment by Profpatsch** [comment]
Update: It works on master.

**Comment by Profpatsch** [comment]
Update: The tests pass on current master.

**Comment by aherrmann** [comment]
> Update: The tests pass on current master.

That's grown to a pretty big diff by now. There's a merge conflict with master now anyway. I've rebased.

**Comment by aherrmann** [comment]
@GrahamcOfBorg build bazel.tests

**Comment by Profpatsch** [comment]
I think ofBorg isn’t listening to Members yet;
@GrahamcOfBorg build bazel.tests

**Comment by Profpatsch** [comment]
Hm, now there’s still one test failing on Darwin:

```
src/main/tools/process-wrapper-legacy.cc:58: "execvp(external/local_config_cc/cc_wrapper.sh, ...)": No such file or directory
INFO: Elapsed time: 3,476s, Critical Path: 0,06s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
builder for '/nix/store/lzii536cvmaix1bnavmwwn5liha3vjl8-bazel-test-cpp.drv' failed with exit code 1
```

**Comment by Profpatsch** [comment]
Ping. So close :)

**Comment by flokli** [comment]
rebased on latest master.

**Comment by flokli** [comment]
@GrahamcOfBorg build bazel.tests

**Pull request merged**